#Project: Exploring eBay Car Sales Data

In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
The data dictionary provided with data is as follows:

dateCrawled - When this ad was first crawled. All field-values are taken from this date.

name - Name of the car.

seller - Whether the seller is private or a dealer.

offerType - The type of listing

price - The price on the ad to sell the car.

abtest - Whether the listing is included in an A/B test.

vehicleType - The vehicle Type.

yearOfRegistration - The year in which the car was first registered.

gearbox - The transmission type.

powerPS - The power of the car in PS.

model - The car model name.

odometer - How many kilometers the car has driven.

monthOfRegistration - The month in which the car was first registered.

fuelType - What type of fuel the car uses.

brand - The brand of the car.

notRepairedDamage - If the car has a damage which is not yet repaired.

dateCreated - The date on which the eBay listing was created.

nrOfPictures - The number of pictures in the ad.

postalCode - The postal code for the location of the vehicle.

lastSeenOnline - When the crawler saw this ad last online.


The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.

1. Import libraries and open the autos dataset

In [46]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [47]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [48]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [49]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

THe dataset contains 20 columns

Some columns have null values but none have more than ~20% null values

The databset has columns that have a mixture of numeric and text values (odometer,model, price)

The name column has values with characters that are separated by underscores and some with /

The dateCrawled Column is in the format YYYY-MM-DD and time in HH:MM:SS

The gearbox column has spelling errors on values automatik and manuel , should be automatic and manual

The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores


Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive

In [50]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [51]:
mapping_dict = {
    'dateCrawled':'date_crawled',
    'name':'name',
    'seller':'seller',
    'offerType':'offer_type',
    'price':'price',
    'abtest':'abtest',
    'vehicleType':'vehicle_type',
    'yearOfRegistration':'registration_year',
    'gearbox':'gearbox',
    'powerPS':'power_ps',
    'model':'model',
    'odometer':'odometer',
    'monthOfRegistration':'registration_month',
    'fuelType':'fuel_type',
    'brand':'brand',
    'notRepairedDamage':'unrepaired_damage',
    'dateCreated':'ad_created',
    'nrOfPictures':'nr_of_pictures',
    'postalCode':'postal_code',
    'lastSeen':'last_seen'
    
     
}


autos.columns = autos.columns.map(mapping_dict)


We have changed the column names from camelcase (Capitalizing the start of every word) to snakecase (using underscores to seperate words)
This improves the naming conversion of our columns and code readability

eg. registration year is more comprehensive than yearOfRegistration

we will use autos.head() to analyse our changes

In [52]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

2. Initial Exploration and Cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. 
Initially we will look for:

1. Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
2. Examples of numeric data stored as text which can be cleaned and converted.

descriptive statistics of all the columns



In [53]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The seller and offer_type columns are candidates to be dropped as it has the same value privat and angebot

Registration year needs more investigation, them maxnimum year is 9999 and the minimum year is 1000

The data in the odometer column is stored as text and needs to be cleaned

The data in the price column is stored as text and needs to be cleaned
(remove non-numeric characters) and (convert the column to a numeric dtype)

Rename odometer to odometer_km using autos.rename()

the data in gearbox column has to be cleaned "manuell" to manual and automaik to automatic


In [54]:
autos["price"] = autos["price"].astype(str).str.replace("$","").str.replace(",","").astype(float)

autos["odometer"] = autos["odometer"].astype(str).str.replace(",","").str.replace("km","").astype(float)


autos.rename({"odometer":"odometer_km"}, axis = 1, inplace = True)
autos.rename(columns={"price":"price_dollar"}, inplace=True)



In [55]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_dollar   abtest vehicle_type  registration_year  \
0  privat    Angebot        5000.0  control          bus               2004   
1  privat    Angebot        8500.0  control    limousine               1997   
2  privat    Angebot        8990.0     test    limousine               2009   
3  privat    Angebot        4350.0  control   kleinwagen               2007   
4  privat    Angebot        1350.0     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove

In [56]:
autos["odometer_km"].unique().shape

autos["odometer_km"].describe()


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

There are 13 unique value in the odometer_km column
The minimum value is 5000 km and the maximum is 150000 km, the mean km is 125732km


In [57]:
autos["odometer_km"].value_counts().sort_index(ascending = False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: odometer_km, dtype: int64

Cars with 150000km have the largest count, whilst the ones with 1000km have the lowest count

In [58]:
autos["price_dollar"].unique().shape



(2357,)

In [59]:
autos["price_dollar"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64

In [60]:
autos["price_dollar"].value_counts().sort_index(ascending = False).head(10)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
Name: price_dollar, dtype: int64

we can observe that there are 2357 values with the minimum price for the vehicle is 0 and the max is 9999999
both values do not make sense. we have cut our dataset by removing outliers. our prices now begin at 3000 to 3890000

In [61]:
max_price = autos["price_dollar"].max()

price_dollar = autos[(autos["price_dollar"] > 0 ) & (autos["price_dollar"] < max_price)]



3. Exploring the date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively


In [62]:
autos[["price_dollar","odometer_km"]].describe()

price_dollar    odometer_km
count  5.000000e+04   50000.000000
mean   9.840044e+03  125732.700000
std    4.811044e+05   40042.211706
min    0.000000e+00    5000.000000
25%    1.100000e+03  125000.000000
50%    2.950000e+03  150000.000000
75%    7.200000e+03  150000.000000
max    1.000000e+08  150000.000000

In [63]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price_dollar   abtest vehicle_type  \
0      privat    Angebot        5000.0  control          bus   
1      privat    Angebot        8500.0  control    limousine   
2      privat    Angebot        8990.0     test    limousine   
3      privat    Angebot        4350.0  control   kleinwagen   
4      privat    Angebot        1350.0     test        kombi   
...       ...        ...           ...      ...          ...   
49995  privat    Angebot       24900.0  control    limousine   
49996  privat    Angebot        1980.0  control       cabrio   
49997  privat    Angebot       13200.0     test       cabrio   
49998  privat    Angebot       22900.0  control        kombi   
49999  privat    Angebot        1250.0  control    limousine   

       registration_year    gearbox  power_ps   model  odometer_km  \
0                   2004    manuell       158  andere     150000.0   
1                   1997  automatik       286     7er     150000.0   
2                   2009    manuell       102    golf      70000.0   
3                   2007  automatik        71  fortwo      70000.0   
4                   2003    manuell         0   focus     150000.0   
...                  ...        ...       ...     ...          ...   
49995               2011  automatik       239      q5     100000.0   
49996               1996    manuell        75   astra     150000.0   
49997               2014  automatik        69     500       5000.0   
49998               2013    manuell       150      a3      40000.0   
49999               1996    manuell       101  vectra     150000.0   

       registration_month fuel_type       brand unrepaired_damage  \
0                       3       lpg     peugeot              nein   
1                       6    benzin         bmw              nein   
2                       7    benzin  volkswagen              nein   
3                       6    benzin       smart              nein   
4                       7    benzin        ford              nein   
...                   ...       ...         ...               ...   
49995                   1    diesel        audi              nein   
49996                   5    benzin        opel              nein   
49997                  11    benzin        fiat              nein   
49998                  11    diesel        audi              nein   
49999                   1    benzin        opel              nein   

                ad_created  nr_of_pictures  postal_code            last_seen  
0      2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54  
1      2016-04-04 00:00:00               0        71034  2016-04-06 14:45:08  
2      2016-03-26 00:00:00               0        35394  2016-04-06 20:15:37  
3      2016-03-12 00:00:00               0        33729  2016-03-15 03:16:28  
4      2016-04-01 00:00:00               0        39218  2016-04-01 14:38:50  
...                    ...             ...          ...                  ...  
49995  2016-03-27 00:00:00               0       

In [64]:
autos[['date_crawled','ad_created','last_seen']] [0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

the first 10 characters represent the day (e.g. 2016-03-12).
To understand the date range, we can extract just the date values,


Extract the date from the date_crawled column

In [65]:
autos['date_crawled'].str[:10]

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 50000, dtype: object

Include the missing values in the distribution and represent the distribution with percentages

In [66]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False)

2016-04-03    0.03868
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-12    0.03678
2016-03-14    0.03662
2016-04-04    0.03652
2016-03-07    0.03596
2016-04-02    0.03540
2016-03-19    0.03490
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-15    0.03398
2016-04-01    0.03380
2016-03-30    0.03362
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-22    0.03294
2016-03-11    0.03248
2016-03-26    0.03248
2016-03-23    0.03238
2016-03-10    0.03212
2016-03-31    0.03192
2016-03-25    0.03174
2016-03-17    0.03152
2016-03-27    0.03104
2016-03-16    0.02950
2016-03-24    0.02910
2016-03-05    0.02538
2016-03-13    0.01556
2016-03-06    0.01394
2016-04-05    0.01310
2016-03-18    0.01306
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

Rank by date in ascending order(earliest to latest)

In [67]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index

<bound method Series.sort_index of 2016-04-03    0.03868
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-12    0.03678
2016-03-14    0.03662
2016-04-04    0.03652
2016-03-07    0.03596
2016-04-02    0.03540
2016-03-19    0.03490
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-15    0.03398
2016-04-01    0.03380
2016-03-30    0.03362
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-22    0.03294
2016-03-11    0.03248
2016-03-26    0.03248
2016-03-23    0.03238
2016-03-10    0.03212
2016-03-31    0.03192
2016-03-25    0.03174
2016-03-17    0.03152
2016-03-27    0.03104
2016-03-16    0.02950
2016-03-24    0.02910
2016-03-05    0.02538
2016-03-13    0.01556
2016-03-06    0.01394
2016-04-05    0.01310
2016-03-18    0.01306
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64>

2016-03-20    0.03782 has the highest percentage of values in the date crawled column
A large number of crawlers were run on the day
2016-04-07    0.00142 has the lowest percentage of crawls

In [68]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index

<bound method Series.sort_index of 2016-04-03    0.03892
2016-03-20    0.03786
2016-03-21    0.03772
2016-04-04    0.03688
2016-03-12    0.03662
               ...   
2015-06-11    0.00002
2016-01-03    0.00002
2016-02-16    0.00002
2016-02-01    0.00002
2016-02-17    0.00002
Name: ad_created, Length: 76, dtype: float64>

Again 2016-03-20    0.03786 has the large number of ads that were created on the day from the website
2016-01-13    0.00002 has the lowest number of ads that were created on the website

In [69]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index

<bound method Series.sort_index of 2016-04-06    0.22100
2016-04-07    0.13092
2016-04-05    0.12428
2016-03-17    0.02792
2016-04-03    0.02536
2016-04-02    0.02490
2016-03-30    0.02484
2016-04-04    0.02462
2016-03-31    0.02384
2016-03-12    0.02382
2016-04-01    0.02310
2016-03-29    0.02234
2016-03-22    0.02158
2016-03-28    0.02086
2016-03-21    0.02074
2016-03-20    0.02070
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-23    0.01858
2016-03-26    0.01696
2016-03-16    0.01644
2016-03-27    0.01602
2016-03-15    0.01588
2016-03-19    0.01574
2016-03-14    0.01280
2016-03-11    0.01252
2016-03-10    0.01076
2016-03-09    0.00986
2016-03-13    0.00898
2016-03-08    0.00760
2016-03-18    0.00742
2016-03-07    0.00536
2016-03-06    0.00442
2016-03-05    0.00108
Name: last_seen, dtype: float64>

2016-04-07    0.13092 highest added by the crawler

In [70]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The maximum year is 9999 which doesnt make any sense as far as present date and old dates
Also the min is 1000 whish also doesnt make sense


Dealing with Incorrect Registration Year Data

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

We will take 1930 as the nacceputeable lowest value because thats where vehicles started getting made and used

we will take 2023 as the highest value that is acceptable as it is the present year. we cant take anything above that as we havent live the time/days

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.



In [71]:
autos["registration_year"] = autos[autos["registration_year"].between(1930,2023)]

In [72]:
autos["registration_year"].describe()

count                   49965
unique                  48181
top       2016-03-08 10:40:35
freq                        3
Name: registration_year, dtype: object

4. Exploring Price by Brand

Explore the unique values in the brand column, and decide on which brands you want to aggregate by

we now select those  brands that have over 5%  of the total values (e.g. > 5%).

In [73]:
car_brands = (autos["brand"].value_counts(normalize=True)[ autos["brand"].value_counts(normalize=True) > .05]).index
car_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [77]:
brand_price_dict = {}


for b in car_brands:
    
    #boolean series
    brand = autos[autos["brand"] == b]
    
    
    # For each brand, get the mean for price_dollar 
    mean_price = brand["price_dollar"].mean()
    

    # Fill the appropriate dictionary with the mean_price
    brand_price_dict[b] =  mean_price
   
    
    
    
    
    
    



In [78]:
brand_price_dict

{'volkswagen': 6384.167399644428,
 'opel': 5106.092657022524,
 'bmw': 8252.918953766808,
 'mercedes_benz': 29511.955428812842,
 'audi': 8965.560354891431,
 'ford': 7105.662546708824}

In above result set, for top 5% brands, we can clearly see that luxury bands (bmw, merc and audi) are costlier than VW, Opel and ford. And if we look at cheaper options then the options are Opel and Ford, however VW is in middle range

5. Storing Aggregate Data in a DataFrame

we aggregated across brands to understand mean price. We observed that in the top 6 brands, there's a distinct price gap.

Audi, BMW and Mercedes Benz are more expensive

Ford and Opel are less expensive

Volkswagen is in between

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

it's difficult to compare more than two aggregate series objects if we want to extend to more columns
we can't compare more than a few rows from each series object
we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. 

In [81]:
brand_milage_dict = {}


for brand in car_brands:
    
    #boolean series
    brands = autos[autos["brand"] == brand]
    
  
    
    # For each brand, get the mean for milage(odometer_km)
    mean_milage = brands["odometer_km"].mean()
    
    
    # Fill the appropriate dictionary with the mean_mileage 
    brand_milage_dict[brand] =  mean_milage
   

In [82]:
brand_milage_dict

{'volkswagen': 128955.27276129878,
 'opel': 129298.66324848929,
 'bmw': 132521.64302818198,
 'mercedes_benz': 130886.14279678918,
 'audi': 129643.9411627364,
 'ford': 124131.93446392642}

In [84]:
mean_price_ser = pd.Series(brand_price_dict).sort_values(ascending = False)
mean_odor_ser = pd.Series(brand_milage_dict).sort_values(ascending = False)

In [86]:
Final_Data_Frame = pd.DataFrame(mean_price_ser, columns = ['mean_price'])
Final_Data_Frame['mean_milage'] = mean_milage

In [87]:
Final_Data_Frame

mean_price    mean_milage
mercedes_benz  29511.955429  124131.934464
audi            8965.560355  124131.934464
bmw             8252.918954  124131.934464
ford            7105.662547  124131.934464
volkswagen      6384.167400  124131.934464
opel            5106.092657  124131.934464

If we look at the above data it seems like odometer (millage) seems uniform however the price varies as per brand. Luxery bards are expensive.